marshal/marshal.go

```diff
@@ -0,0 +1,190 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"golang.org/x/exp/maps"
+)
+
+const forceSendFieldName = "ForceSendFields"
+
+// Marshal returns a JSON encoding of the given object. Included fields:
+// - non-empty value
+// - a basic type whose field's name is present in forceSendFields
+// Embedded structs are still considered a separate struct. ForceSendFields
+// in an embedded struct only impact the fields of the embedded struct.
+// Conversely, an embedded struct is not impacted by the ForceSendFields
+// of the struct containing it.
+func Marshal(object any) ([]byte, error) {
+	dataMap, err := structAsMap(object)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(dataMap)
+}
+
+// Converts the object to a map
+func structAsMap(object any) (map[string]any, error) {
+	result := make(map[string]any)
+
+	// If the object is nil or a pointer to nil, don't do anything
+	if object == nil || (reflect.ValueOf(object).Kind() == reflect.Ptr && reflect.ValueOf(object).IsNil()) {
+		return result, nil
+	}
+
+	value := reflect.ValueOf(object)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	includeFields, err := getForceSendFields(object, getTypeName(objectType))
+
+	if err != nil {
+		return nil, err
+	}
+
+	fields := getTypeFields(objectType)
+
+	for i := 0; i < value.NumField(); i++ {
+		jsonTag := fields[i].JsonTag
+		tag, err := parseJSONTag(jsonTag)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
There are 2 things happening here that I think need to be changed:
* The JSON tag is stored as a string on the field and is then parsed again (albeit cached). Why not store it on the cached type field in the first place?
* Iteration happens on `value.NumField()` as opposed to `len(fields)`. Instead, the cached type field can include its field index in the struct type, so that you can range over `fields` directly.